core/src/main/java/org/apache/calcite/sql/SqlCall.java

```diff
@@ -118,7 +118,7 @@ public int operandCount() {
     final SqlDialect dialect = writer.getDialect();
     if (leftPrec > operator.getLeftPrec()
         || (operator.getRightPrec() <= rightPrec && (rightPrec != 0))
-        || writer.isAlwaysUseParentheses() && isA(SqlKind.EXPRESSION)) {
+        || writer.isAlwaysUseParentheses() && isA(SqlKind.EXPRESSION) && !writer.inWithBody()) {
```

Comment:
Why must you add this condition? I know `testWithUnion` will fail without this, however, it's surprising that `SqlKind.SET_QUERY` is a `SqlKind.EXPRESSION` while `SELECT`/`ORDER_BY`/`WITH`/`JOIN` etc. are not.

Reply:
Yes, it's very strange that `SqlKind.EXPRESSION` contains UNION, and according to declaration of `SqlKind.EXPRESSION`, it should contains `SELECT/ORDER_BY/WITH/JOIN` etc, and not contains `UNION`, but when i debug, something is different, you can not find `SELECT/ORDER_BY/WITH/JOIN` in `SqlKind.EXPRESSION`, and `UNION` will appear. Let's back to the topic, might putting `inWithBody()` into `SqlWriter#isAlwaysUseParentheses ` will be better?

Reply:
`SqlKind.EXPRESSION` uses `EnumSet.complementOf`, hence its document and its result is consistent. The question is "should SET_QUERY be a EXPRESSION"?

Reply:
IMHO, SET_QUERY should not be a EXPRESSION.

Reply:
> IMHO, SET_QUERY should not be a EXPRESSION.

Then you can add `SET_QUERY` to `SqlKind.EXPRESSION`, and you do not need to add this condition `!writer.inWithBody()` here anymore.

Reply:
You are right, exclude "SET_QUERY" from EXPRESSION will be better, but many test cases may fail if do that, I will improve it in next commit.